docs/configuration.rst

Configuration file
===================

The configuration file that is to be provided to the interfaces
specify file paths to policy models and stocks, in addition to
detailed parameters of the tree search.

Simple usage
------------

Let's say you have a

    * A trained Keras expansion model that is called `uspto_expansion.onnx`
    * A library of unique templates called `uspto_templates.csv.gz`
    * A stock file in HDF5 format, called `zinc_stock.hdf5`

(this could have been created by the ``smiles2stock`` tool, see :doc:`here <stocks>`)

then to use them in the tree search, create a file called `config.yml` that has the following content

.. code-block:: yaml

    expansion:
      full:
        - uspto_expansion.onnx
        - uspto_templates.csv.gz
    stock:
      zinc: zinc_stock.hdf5


Advanced usage
--------------

A more detailed configuration file is shown below

.. code-block:: yaml

    search:
      algorithm: mcts
      algorithm_config:
        C: 1.4
        default_prior: 0.5
        use_prior: True
        prune_cycles_in_search: True
        search_reward: state score
      max_transforms: 6
      iteration_limit: 100
      return_first: false
      time_limit: 120
      exclude_target_from_stock: True 
    expansion:
      my_policy:
        type: template-based
        model: /path/to/keras/model/weights.hdf5
        template: /path/to/hdf5/templates.hdf5
        template_column: retro_template
        cutoff_cumulative: 0.995
        cutoff_number: 50
        use_rdchiral: True
        use_remote_models: False
      my_full_policy:
        - /path/to/keras/model/weights.hdf5
        - /path/to/hdf5/templates.hdf5
    filter:
      uspto:
        type: quick-filter
        model: /path/to/keras/model/weights.hdf5
        exclude_from_policy: rc
        filter_cutoff: 0.05
        use_remote_models: False
      uspto_full: /path/to/keras/model/weights.hdf5
    stock:
      buyables:
        type: inchiset
        path: /path/to/stock1.hdf5
      emolecules: /path/to/stock1.hdf5

The (expansion) policy models are specified using two files
    * a checkpoint files from Keras in ONNX or hdf5 format,
    * a HDF5 or a CSV file containing templates.

A key like ``my_policy`` should be set and the configuration contains ``type``, ``model`` and ``template`` that must be provided. 
If the other settings are not assigned, their default values are taken. 
The policy ``my_full_policy`` exemplifies a short-cut to the template-based expansion model when no other settings need to be 
provided, only the ``model`` and ``templates`` can be provided. The default settings will be taken in this case.

The template file should be readable by ``pandas`` using  the ``table`` key and the ``retro_template`` column.
A policy can then be selected using the provided key, like ``my_policy`` in the above example.

The filter policy model is specified using a single checkpoint file.
Any key like ``uspto`` can be set. The settings contains ``type`` and ``model`` that must be provided. If the other
settings are not assigned, their default values are taken. 
The filter ``uspto_full`` exemplifies a short-cut to the quick-filter model when no other settings need to be 
provided, only the ``model`` can be provided. The default settings will be taken in this case.


The stock files can be
     * HDF5 files with the ``table`` key an the ``inchi_key`` column.
     * A CSV file with a ``inchi_key`` column
     * A text file a single column

In all cases, the column should contain pre-computed inchi keys of the molecules.
The stocks can be set using any key, like ``buyables`` or ``emolecules`` in the above example.
The ``type`` and ``path`` parameters can also be set along with other parameters.
If no other settings need to be provided, only the ``path`` can be provided, whereby it will be 
treated as a 
short-cut to the inchiset class..

The values in the ``search`` sections are optional, and if missing, default values are considered. These
values can also be taken from environment variables. An example of this can be seen as below:

.. code-block:: yaml

    search:
      iteration_limit: ${ITERATION_LIMIT}
      algorithm_config:
        C: ${C}
      time_limit: ${TIME_LIMIT}
      max_transforms: ${MAX_TRANSFORMS}

These are the available ``search`` settings. The ``algorithm_config`` refers to MCTS settings:

============================================ ============== ===========
Setting                                      Default value  Description
============================================ ============== ===========
algorithm                                    mcts           The search algorithm. Can be set to `package.module.ClassName` to use a custom search method.
algorithm_config: C                          1.4            The C value used to balance exploitation and exploration in the upper confidence bound score of the nodes.
algorithm_config: default_prior              0.5            The prior that is used if policy-provided priors are not used.
algorithm_config: use_prior                  True           If True, priors from the policy is used instead of the `default_prior`.
algorithm_config: prune_cycles_in_search     True           If True, prevents the MCTS from creating cycles by recreating previously seen molecules when it is expanded.
algorithm_config: search_reward              state score    The scoring used for the MCTS search algorithm.
algorithm_config: immediate_instantiation    []             list of expansion policies for which the MCTS algorithm immediately instantiate the children node upon expansion
algorithm_config: mcts_grouping              -              if is partial or full the MCTS algorithm will group expansions that produce the same state. If ``partial`` is used the equality will only be determined based on the expandable molecules, whereas ``full`` will check all molecules.
max_transforms                               6              The maximum depth of the search tree.
iteration_limit                              100            The maximum number of iterations for the tree search.
time_limit                                   120            The maximum number of seconds to complete the tree search.
return_first                                 False          If True, the tree search will be terminated as soon as one solution is found.
exclude_target_from_stock                    True           If True, the target is in stock will be broken down.
============================================ ============== ===========


The ``post_processing`` settings are:

============================================ ============== ===========
Setting                                      Default value  Description
============================================ ============== ===========
min_routes                                   5              The minumum number of routes to extract if ``all_routes`` is not set.
max_routes                                   25             The maximum number of routes to extract if ``all_routes`` is not set.
all_routes                                   False          If True, will extract all solved routes.
route_distance_model                         N/A            If set, will load the quick route distance model from this checkpoint file.
route_scorer                                 state score    The scoring for routes when extracting them in the post processing step.
============================================ ============== ===========


The ``expansion`` settings are for template-based models:

============================================ ============== ===========
Setting                                      Default value  Description
============================================ ============== ===========
template_column                              retro_template The column in the template file that contains the templates.
cutoff_cumulative                            0.995          The accumulative probability of the suggested templates is capped at this value. All other templates above this threshold are discarded.
cutoff_number                                50             The maximum number of templates that will be returned from the expansion policy.
use_rdchiral                                 True           If True, will apply templates with RDChiral, otherwise RDKit will be used.
use_remote_models                            False          If True, will try to connect to remote Tensorflow servers.
rescale_prior                                False          If True, will apply a softmax function to the priors.
============================================ ============== ===========


The ``filter`` settings are for quick-filter models:

============================================ ============== ===========
Setting                                      Default value  Description
============================================ ============== ===========
exclude_from_policy                          []             The list of names of the filter policies to exclude.
filter_cutoff                                0.05           The cut-off for the quick-filter policy.
use_remote_models                            False          If True, will try to connect to remote Tensorflow servers.
============================================ ============== ===========


docs/source/configuration.rst

Configuration file
===================

The configuration file that is to be provided to the interfaces 
specify file paths to policy models and stocks, in addition to 
detailed parameters of the tree search. 

Simple usage
------------

Let say you have a 

    * A trained Keras expansion model that is called `full_uspto_model.hdf5`
    * A library of unique templates called `full_uspto_templates.hdf5`

(these could have been created by the training tools, see :doc:`here <training>`)

    * A stock file in HDF5 format, called `zinc_stock.hdf5`

(this could have been created by the ``smiles2stock`` tool, see :doc:`here <stocks>`)

then to use them in the tree search, create a file called `config.yml` that has the following content 

.. code-block:: yaml

    policy:
      files:
        full_uspto:
          - full_uspto_model.hdf5
          - full_uspto_templates.hdf5
    stock:
      files:
        zinc: zinc_stock.hdf5


Advanced usage
--------------

A more detailed configuration file is shown below

.. code-block:: yaml

    properties:
      iteration_limit: 100
      return_first: false
      time_limit: 120
      C: 1.4
      cutoff_cumulative: 0.995
      cutoff_number: 50
      max_transforms: 6
    policy:
      files:
        my_policy:
          - /path/to/keras/model/weights.hdf5
          - /path/to/hdf5/templates.hdf5
    filter:
      files:
        my_policy: /path/to/keras/model/weights.hdf5
    stock:
      files:
        stock1: /path/to/stock1.hdf5
        stock2: /path/to/stock1.hdf5

The (expansion) policy models are specified using two files
    * a checkpoint files from Keras in hdf5 format,
    * a HDF5 file containing templates. 

The filter policy model is specified using a single checkpoint file from Keras in hdf5 format.
    
The template file should be readable by ``pandas`` using  the ``table`` key and the ``retro_template`` column. 
A policy can then be selected using the provided key, like ``my_policy`` in the above example. 

The stock files should be HDF5 files with the ``table`` key an the ``inchi_key`` column. The column
should contain pre-computed inchi keys of the molecules. 
The stocks can then be selected using the provided key, like ``stock1`` or ``stock2`` in the above example.

The values in the ``properties`` sections are optional, and if missing, default values are provided.

These are the available properties:

========================= ============== ===========
Property                  Default value  Description
========================= ============== ===========
C                         1.4            The C value used to balance exploitation and exploration in the upper confidence bound score of the nodes.
cutoff_cumulative         0.995          The accumulative probability of the suggested templates is capped at this value. All other templates above this threshold are discarded. 
cutoff_number             50             The maximum number of templates that will be returned from the expansion policy.
max_transforms            6              The maximum depth of the search tree
default_prior             0.5            The prior that is used if policy-provided priors are not used
use_prior                 True           If true, priors from the policy is used instead of the `default_prior`
return_first              False          If true, the tree search will be terminated as soon as one solution is found
iteration_limit           100            The maximum number of iterations for the tree search
time_limit                120            The maximum number of seconds to complete the tree search
exclude_target_from_stock True           If the target is in stock it will be broken down if this property is True
template_column           retro_template the column in the template file that contains the templates
filter_cutoff             0.05           the cut-off for the filter policy
prune_cycles_in_search    True           prevents the MCTS from creating cycles by recreating the parent of a molecule when it is expanded
========================= ============== ===========

CHANGELOG.md
# CHANGELOG

## Version 4.0.0 2023-11-30

### Features

- `Configuration` now supports a `rescale_prior` property which rescales the priors in `TemplateBasedExpansionStrategy`.
- Functionality of `ScorerCollection` has been extended.
- Pricing is now supported in `InMemoryInchiKeyQuery`.
- Reward scorer has been added to Configuration scorers as `search_reward` item.
- `MaxTransformScorerer` and `FractionInStockScorer` have been created to separate scores in `StateScorer`.
- Reaction routes are scored with all reward scorers after the search is complete.
- `StockAvailabilityScorer` and `ReactionClassMembershipScorer` have been added to the scorers.
- Atom mapping existing in target molecule can be inherited.
- A caching feature has been add to the expansion strategies.
- Degenerate states can now be grouped in the MCTS algorithm.
- `ChemformerBasedExpansionStrategy` and `ModelZooExpansionStrategy` can be found under plugins and used as additional expansion strategies.
- Support for the stereocenter model has been added with to support chiral fingerprints for molecules and reactions.
- The `Configuration` format has been entirely revamped to a more easy-to-use format.
- A `MolbloomFilterQuery` has been created as a stock query class.

### Trivial changes

- Python version requirements have been updated to versions 3.9 - 3.11.
- `MoleculeCost` has been moved from aizynthfinder.context.cost.collection to the Retro* package.

### Deprecations

- Graphviz has been removed from aizynthfinder.utils.image.
- `Reaction` class has been removed from aizynthfinder.chem.reaction.
- aizynthfinder.context.cost package has been removed.

### Bug-fixes

- Fixed an issue with `max_transforms` to ensure only the given number of maximum depth is considered.
- Pinned Jupyter notebook version to ^6.5.3 to avoid errors when displaying widgets.
- Rollout child has been removed from the MCTS search algorithm.

## Version 3.7.0 2023-06-01 (2023-04-11)

### Features

- Environment variables can be loaded from YAML file.
- Restart feature in AiZynth CLI.
- Use ONNX models for expansion and filtering.
- Default filter is now loaded automatically in AiZynth CLI.

### Trivial changes

- Using isort to sort the imports of packages.
- Tensorflow is now an optional dependency.
- more-itertools dependencies removed.

### Breaking changes

- training modules has now been removed.

### Bug-fixes

- Fixing an issue when concatenating JSON data files.

## Version 3.6.0 2022-11-28 (2022-11-25)

### Features

- aizynthfinder can now be installed as a pure-python package
- ReactionTree now store at what iteration the route was created
- ReactionTree now supports a metadata property
- MCTS nodes populate a property that stores at what iteration the route was created
- aizynthcli now outputs stock information and route metadata and scores
- aizynthcli now has better error handling of invalid SMILES input
- Graphviz dependency for route drawing is removed, pure python implementation is used instead
- Dependencies reworked so that a minimal package can be installed
- Extra dependencies are related to route distances, clustering, training, MongoDB and external models
- Downloaded files are now the latest USPTO and Ringbreaker models

### Trivial changes

- RDKit and route-distances package now installed from pypi
- Documentation updated and extended

## Version 3.5.0 2022-11-28 (2022-07-21)

### Features

- Atom-mapping is tracked from product to reactant
- Support loading of template library from (gzipped) CSV file
- Support of saving aizynthcli output to (gzipped) JSON file
- AiZynthExpander now tracks non-applicable templates

### Bug-fixes
- Fixed failing test case

### Trival changes
- Silent progress bar when utilizing local Keras model

## Version 3.4.0 2022-04-28

### Features

- Gracefully fail predictions with aizynthcli if target is unsanitizable (Github issue 66)

### Trivial changes
- Update version of Sphinx, scikit-learn and Black
- Remove usage of depracted Pandas function append (Github issue 63)
- Correct documentation of aizynthcli (Github issue 67)

## Version 3.3.1 2022-03-09

### Trivial changes

- Updated pinned versions of route_distances, jinja2 and tensorflow

## Version 3.3.0 2022-02-24

### Features

- Support for Retro* tree search
- Support for breadth-first exhaustive tree search
- Support for depth-first proof-number tree search
- Possible to save concatenated reaction trees to separate file

### Bugfixes

- RouteCostScorer fix for rare routes

## Version 3.2.0 2022-02-24

### Features

- Profiling feature enabled in search trees
- New, customizable configuration of training pre-processing tools
- Generic post-processing support in aizynthcli
- Introduce short aliases for filter policies
- Reaction shape support in GraphViz visualisation of routes

## Version 3.1.0 2021-12-21

### Features

- ReactantsCountFiler (Github issue 42) to filter reactions with incompatible number of reactants
- ForwardRegenerationFiler to filter reactions where the forward reaction is incompatible
- Possible to skip quick Keras filter for specific policies
- Possible to select more than one policy in the GUI application
- Reaction classes has a hash function
- Possible to extract sub trees from ReactionTree objects
- RDKit can be used instead of RDChiral for expansions

### Bugfixes

- Possible to use more than depth 6 in the GUI application
- Fix failure in MctsNode class when expansion policy return no molecules

### Trivial changes

- Update type hints to be compatible with latest numpy release
- Update route-distances dependency

## Version 3.0.0  2021-07-26

### Features

- Generalized expansion and filter policies - policies no longer need to be Keras models
- Updated image generation for synthesis routes
- Improved code to prune "regeneration reactions" in MCTS (Github issue 38)
- Reactants are sorted in output from ReactionTree class
- New arguments to the "aizynthcli" tool

aizynthfinder/context/config.py
""" Module containing a class for encapsulating the settings of the tree search
"""
from __future__ import annotations

import os
import re
from dataclasses import dataclass, field
from typing import TYPE_CHECKING

import yaml
from aizynthfinder.context.policy import ExpansionPolicy, FilterPolicy
from aizynthfinder.context.scoring import ScorerCollection
from aizynthfinder.context.stock import Stock
from aizynthfinder.utils.logging import logger

if TYPE_CHECKING:
    from aizynthfinder.utils.type_utils import Any, Dict, List, Optional, StrDict, Union


@dataclass
class _PostprocessingConfiguration:
    min_routes: int = 5
    max_routes: int = 25
    all_routes: bool = False
    route_distance_model: Optional[str] = None
    route_scorer: str = "state score"


@dataclass
class _SearchConfiguration:
    algorithm: str = "mcts"
    algorithm_config: Dict[str, Any] = field(
        default_factory=lambda: {
            "C": 1.4,
            "default_prior": 0.5,
            "use_prior": True,
            "prune_cycles_in_search": True,
            "search_reward": "state score",
            "immediate_instantiation": (),
            "mcts_grouping": None,
        }
    )
    max_transforms: int = 6
    iteration_limit: int = 100
    time_limit: int = 120
    return_first: bool = False
    exclude_target_from_stock: bool = True
    break_bonds: List[List[int]] = field(default_factory=list)
    freeze_bonds: List[List[int]] = field(default_factory=list)
    break_bonds_operator: str = "and"


@dataclass
class Configuration:
    """
    Encapsulating the settings of the tree search, including the policy,
    the stock, the loaded scorers and various parameters.
    """

    search: _SearchConfiguration = field(default_factory=_SearchConfiguration)
    post_processing: _PostprocessingConfiguration = field(
        default_factory=_PostprocessingConfiguration
    )
    stock: Stock = field(init=False)
    expansion_policy: ExpansionPolicy = field(init=False)
    filter_policy: FilterPolicy = field(init=False)
    scorers: ScorerCollection = field(init=False)

    def __post_init__(self) -> None:
        self.stock = Stock()
        self.expansion_policy = ExpansionPolicy(self)
        self.filter_policy = FilterPolicy(self)
        self.scorers = ScorerCollection(self)
        self._logger = logger()

    def __eq__(self, other: Any) -> bool:
        if not isinstance(other, Configuration):
            return False
        for key, setting in vars(self).items():
            if isinstance(setting, (int, float, str, bool, list)):
                if (
                    vars(self)[key] != vars(other)[key]
                    or self.search != other.search
                    or self.post_processing != other.post_processing
                ):
                    return False
        return True

    @classmethod
    def from_dict(cls, source: StrDict) -> "Configuration":
        """
        Loads a configuration from a dictionary structure.
        The parameters not set in the dictionary are taken from the default values.
        The policies and stocks specified are directly loaded.

        :param source: the dictionary source
        :return: a Configuration object with settings from the source
        """
        expansion_config = source.pop("expansion", {})
        filter_config = source.pop("filter", {})
        stock_config = source.pop("stock", {})
        scorer_config = source.pop("scorer", {})

        config_obj = Configuration()
        config_obj._update_from_config(dict(source))

        config_obj.expansion_policy.load_from_config(**expansion_config)
        config_obj.filter_policy.load_from_config(**filter_config)
        config_obj.stock.load_from_config(**stock_config)
        config_obj.scorers.create_default_scorers()
        config_obj.scorers.load_from_config(**scorer_config)

        return config_obj

    @classmethod
    def from_file(cls, filename: str) -> "Configuration":
        """
        Loads a configuration from a yaml file.
        The parameters not set in the yaml file are taken from the default values.
        The policies and stocks specified in the yaml file are directly loaded.
        The parameters in the yaml file may also contain environment variables as
        values.

        :param filename: the path to a yaml file
        :return: a Configuration object with settings from the yaml file
        :raises:
            ValueError: if parameter's value expects an environment variable that
                does not exist in the current environment
        """
        with open(filename, "r") as fileobj:
            txt = fileobj.read()
        environ_var = re.findall(r"\$\{.+?\}", txt)
        for item in environ_var:
            if item[2:-1] not in os.environ:
                raise ValueError(f"'{item[2:-1]}' not in environment variables")
            txt = txt.replace(item, os.environ[item[2:-1]])
        _config = yaml.load(txt, Loader=yaml.SafeLoader)
        return Configuration.from_dict(_config)

    def _update_from_config(self, config: StrDict) -> None:
        self.post_processing = _PostprocessingConfiguration(
            **config.pop("post_processing", {})
        )

        search_config = config.pop("search", {})
        for setting, value in search_config.items():
            if value is None:
                continue
            if not hasattr(self.search, setting):
                raise AttributeError(f"Could not find attribute to set: {setting}")
            if setting.endswith("_bonds"):
                if not isinstance(value, list):
                    raise ValueError("Bond settings need to be lists")
                value = _handle_bond_pair_tuples(value) if value else []
            if setting == "algorithm_config":
                if not isinstance(value, dict):
                    raise ValueError("algorithm_config settings need to be dictionary")
                self.search.algorithm_config.update(value)
            else:
                setattr(self.search, setting, value)


def _handle_bond_pair_tuples(bonds: List[List[int]]) -> List[List[int]]:
    if not all(len(bond_pair) == 2 for bond_pair in bonds):
        raise ValueError("Lists of bond pairs to break/freeze should be of length 2")
    return [bond_pair[:2] for bond_pair in bonds]
